File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/roadrunner/tuning/ManualFeedbackTuner.java
package org.firstinspires.ftc.teamcode.roadrunner.tuning;

import com.acmerobotics.roadrunner.Pose2d;
// TODO: remove Actions from the core module?
import com.acmerobotics.roadrunner.ftc.Actions;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

import org.firstinspires.ftc.teamcode.roadrunner.MecanumDrive;

public final class ManualFeedbackTuner extends LinearOpMode {
    public static double DISTANCE = 64;

    @Override
    public void runOpMode() throws InterruptedException {
        MecanumDrive drive = new MecanumDrive(hardwareMap, new Pose2d(0, 0, 0));

        waitForStart();

        while (opModeIsActive()) {
            Actions.runBlocking(
                drive.actionBuilder(drive.pose)
                        .lineToX(DISTANCE)
                        .lineToX(0)
                        .build());
        }
    }
}

File: TeamCode/src/main/main/java/org/firstinspires/ftc/teamcode/drivetrain/tuning/ManualFeedbackTuner.kt
package org.firstinspires.ftc.teamcode.drivetrain.tuning

import com.acmerobotics.roadrunner.Pose2d
import com.acmerobotics.roadrunner.ftc.runBlocking
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.firstinspires.ftc.teamcode.drivetrain.MecanumDrive
import org.firstinspires.ftc.teamcode.drivetrain.ThreeDeadWheelLocalizer
import java.lang.RuntimeException

class ManualFeedbackTuner : LinearOpMode() {
    @kotlin.Throws(java.lang.InterruptedException::class)
    override fun runOpMode() {
        val drive: MecanumDrive = MecanumDrive(hardwareMap, Pose2d(0.0, 0.0, 0.0))

        if (ThreeDeadWheelLocalizer.PARAMS.perpXTicks.toInt() == 0 && ThreeDeadWheelLocalizer.PARAMS.par0YTicks.toInt() == 0 && ThreeDeadWheelLocalizer.PARAMS.par1YTicks.toInt() == 1) {
            throw RuntimeException("Odometry wheel locations not set! Run AngularRampLogger to tune them.")
        }

        waitForStart()

        while (opModeIsActive()) {
            runBlocking(
                drive.actionBuilder(Pose2d(0.0, 0.0, 0.0)).lineToX(DISTANCE).lineToX(0.0).build()
            )
        }
    }

    companion object {
        var DISTANCE: kotlin.Double = 64.0
    }
}
